Add AES-XTS

Patater opened the pull request:
Add the XTS block cipher mode to the library. Continued from #414.

TODO
- ~~Decide if we want a new public API for multiply by _x_ GF(128), or turn this into a static function inside aes.c (the only place that currently uses it)~~ Turned into a static function inside aes.c
- ~~Resolve test issues for "too large" sectors.~~ The test still fails with a segfault on NULL access (not garbage), so we aren't letting bugs slip through. Error reporting could be improved here, though (maybe later).
- ~~Refactor `aes_encrypt_xts` and `aes_decrypt_xts` test functions (in `tests/suites/test_suite_aes.function`) to avoid duplicate code, either by combining the functions or adding a shared helper function.~~ We just live with duplicated code in our tests.

Patater commented:
@R030t1 Thanks for your input. Another way to do this would be with a double-sized key, with key1 (main operation) followed by key2 (tweak computation). The tricky part of the XTS API in this PR as currently written is that `ctx_t` should always be used in encrypt mode (`mbedtls_aes_setkey_enc()`), while `ctx` is used in encrypt or decrypt mode (`mbedtls_aes_setkey_enc()` or `mbedtls_aes_setkey_dec()`). This is a gotcha I'd ideally rather not have, but it may be acceptable since the AES API is fairly low-level nuts and bolts. We could instead create a new `mbedtls_xts_setkey_enc()` that takes a double-sized key and properly initializes both AES contexts, but that may be overkill for a "pure mechanism" API like AES.

The cipher APIs are written for a higher level of abstraction than the AES APIs. So, as currently written in the PR, XTS when used from the cipher API takes double-sized keys. XTS is easier to use with the cipher API, but still has some gotchas at the lower AES API level. This means there is some difference in how to use XTS via cipher APIs vs AES APIs, and potential confusion because the API works differently in cipher APIs vs AES APIs. Whether that's acceptable or not needs some more thought.

yanesca commented on include/mbedtls/aes.h:
```diff
@@ -329,6 +329,57 @@ int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
+ *
+ *             The key for tweak computation is separate from the key for AES
+ *             block encryption. So, two AES contexts must be supplied. XXX
+ *             Design review: It's a bit strange to have two AES ctxs.
```
Unfortunately the extra key does not fit in the existing context in all cases so we can't store the second key there. Adding a pointer would be a dead weight in all the other cases, so I can see why you didn't go with it. I feel that adding a new context adds complexity and it is something that I too would be reluctant to do.

However I think that there is less opportunity to misuse the API if there is a dedicated XTS context and in this instance I would vote for defensive coding over conceptual cleanliness.

I was thinking something like this:
``` C
typedef struct
{
    mbedtls_aes_context sub_ctx[2];
}
mbedtls_aes_xts_context;
```

yanesca replied:
That is even better, good idea!

Patater replied:
@yanesca Would you also recommend new `mbedtls_aes_xts_setkey_enc(mbedtls_aes_xts_context* ctx)` and `mbedtls_aes_xts_setkey_dec(mbedtls_aes_xts_context* ctx)` functions?

yanesca replied:
Yes, I think that would be the cleanest solution. Especially if we add the usual `_init()` and `_free()` along with them.

Patater replied:
I've changed the AES XTS adding commit to introduce a new context type and the associated functions.

yanesca replied:
Thanks! The API looks good to me now.

yanesca commented on include/mbedtls/aes.h:
```diff
@@ -329,6 +329,57 @@ int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
+                           mbedtls_aes_context *ctx_t,
+                           int mode,
+                           size_t length,
+                           const unsigned char *sector,
```
Since we know that this is 16 byte long, writing
``` diff
- const unsigned char *sector
+ const unsigned char sector[16]
```
would be slightly more expressive.

Patater replied:
I will change. The parameter array length is not enforced, so this is functionally not any different, but it does tell the programmer what we expect more clearly than just the documentation.

yanesca commented on include/mbedtls/aes.h:
```diff
@@ -329,6 +329,57 @@ int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
+ *             pointer in the ctx for the tweak's ctx, or could make a new ctx
+ *             for xts operations.
+ *
+ *             AES-XTS encrypts or decrypts blocks based on their location as
```
The interface in the standard allows random access to a single AES blocks inside of a data unit. It seems to me that this can be useful depending on the hardware and the application.

Do you think it would worth adding this capability to the API?

Patater replied:
I considered this. Supporting sub-data-unit operations complicates both the API and the implementation. XTS is typically used with whole data units, where the data unit is the size of the block device block. There may be some optimization for small reads to be gained if we support sub-data-unit access, as the entire device block doesn't need decrypting.

If we need sub-data-unit operations, I would propose we add it with a new API at a later time, and keep this simpler API around both as a backwards compatibility wrapper and as an easier to use version of the API.

Patater commented on include/mbedtls/aes.h:
```diff
@@ -329,6 +329,57 @@ int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
+ * \param length       The length of both an entire sector and the input data
+ *                     in bytes. The length must be at least 16 bytes. The
+ *                     length does not need to be a multiple of 16 bytes.
+ * \param sector       The sector address in byte array form. The least
```
Yeah, I will update this to match the terminology used in the standard.

Patater commented on tests/suites/test_suite_aes.function:
```diff
@@ -151,6 +151,157 @@ exit:
+    unsigned char iv_str[100]  = { 0, };
+    unsigned char src_str[100] = { 0, };
+    unsigned char dst_str[100] = { 0, };
+    unsigned char output[100]  = { 0, };
```
This should also be stack friendly, but currently isn't. Change XEX test cases to match XTS test case style.

yanesca commented on library/gf128mul.c:
```diff
@@ -0,0 +1,78 @@
+
+
+/*
+ * This function multiply a field element by x, by x^4 and by x^8
```
I am sorry, but I don't understand this sentence. The sentence "This function multiplies a field element by x." would make sense, but I can't figure out how "x^4" and "x^8" come to the picture. Could you please explain?

Patater replied:
Maybe the x^4 and x^8 are from the irreducible polynomial? I'm not sure either. @Aorimn ?

yanesca replied:
That is not likely to be the case, because the irreducible polynomial here is x^128+x^7+x^2+x+1.

yanesca commented on library/gf128mul.c:
```diff
@@ -0,0 +1,78 @@
+
+
+/* Jump table for not having ifs */
+static const uint16_t gf128mul_table_bbe[2] = { 0x00, 0x87 };
```
I think `uint8_t` should suffice here.

Patater replied:
Fixed.

It's possibly slower to use `uint8_t` on x86 and ARM due to how operations on 8-bit values work on those platforms, but given that the library doesn't use `uint_fast8_t` yet, the original author may have elected to use `uint16_t` as an optimization. Use of `uint16_t` in that manner is not portable or readable, so I've changed this to use `uint8_t` even though that's potentially slower than `uint_fast8_t`.

yanesca replied:
If it is indeed faster, then I would prefer to have a comment explaining the choice of type rather than slowing down the implementation.

Patater replied:
Fixed by removing the jump table

yanesca commented on tests/suites/test_suite_aes.function:
```diff
@@ -151,6 +151,157 @@ exit:
+void aes_decrypt_xts( char *hex_key_string, char *hex_data_unit_string,
+                      char *hex_dst_string, char *hex_src_string )
+{
+    /* XXX Code Review: I don't really like all this duplicated code between
```
I too like the idea of unifying them.

Patater replied:
Yeah, but I believe we should optimize for clarity when a test fails over the maintainability of the test. Adding a shared helper function is great for the maintainability of this code, but if that makes it harder to see where an error is coming from when a test fails, I'd rather pay the maintenance cost: it'd be worth it.

Would you agree? If so, I can delete this junk.

yanesca replied:
When a test fails we know which test case failed and the printed info is the same whether we are using a helper function or not. If we don't know what is going on based on what is printed out, then we need to get our hands dirty anyway. I don't know how much reducing call depth helps when we are inserting printf's in the code or running a debugger anyway.

In summary I am still leaning towards unifying them, but don't consider this as a blocker.

Patater replied:
I'll leave them separate for now. I've removed my XXX comment requesting feedback. I think getting a line number from an assertion failing in a helper function is not so great.

yanesca commented on tests/suites/test_suite_aes.function:
```diff
@@ -151,6 +151,157 @@ exit:
+    mbedtls_aes_xts_init( &ctx );
+
+    data_unit = unhexify_alloc( hex_data_unit_string, &data_unit_len );
+    TEST_ASSERT( data_unit_len == AES_BLOCK_SIZE );
```
If this fails then the test vector is wrong not the library. Do we need to test this?

gilles-peskine-arm replied:
It's cheap and makes it easier to debug a mistake in the test data. I wouldn't insist on it, since it's present I prefer to keep it. This applies to other similar cases.

Patater replied:
No change intended here

yanesca commented on tests/suites/test_suite_aes.function:
```diff
@@ -151,6 +151,157 @@ exit:
+    TEST_ASSERT( data_unit_len == AES_BLOCK_SIZE );
+
+    key = unhexify_alloc( hex_key_string, &key_len );
+    TEST_ASSERT( key_len % 2 == 0 );
```
If this fails then the test vector is wrong not the library. Do we need to test this?

Patater replied:
No change intended here

yanesca commented on tests/suites/test_suite_aes.function:
```diff
@@ -151,6 +151,157 @@ exit:
+
+    src = unhexify_alloc( hex_src_string, &src_len );
+    dst = unhexify_alloc( hex_dst_string, &dst_len );
+    TEST_ASSERT( src_len == dst_len );
```
If this fails then the test vector is wrong not the library. Do we need to test this?

R030t1 replied:
I would be worried about having improperly described tests generate results which claim to pass. If lucky these failures would mean the library could only work with itself, but it is possible (though very unlikely) that a weakness could generate code compatible with other implementations that is weaker in some way. The biggest concern would be key generation, which this is not.

So in this case the test might be redundant. In general I am not sure it is a good idea to discourage test sanity checks.

Patater replied:
I have this test present to catch typos in the test vectors file. For instance, if you copy paste everything but the last little bit of the `src`, we'd catch that here.

yanesca commented on tests/suites/test_suite_aes.function:
```diff
@@ -151,6 +151,157 @@ exit:
+/* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_CIPHER_MODE_XTS */
+void aes_decrypt_xex( char *hex_key_string, char *hex_iv_string,
```
The `_xex` and `_xts` functions in this file are practically the same, the only thing that seems to be different is the allocation strategy. Why do we need them both?

Patater replied:
Another difference is that one calls xex and the other calls xts. Is that worth having both for?

OK, so another reason for having a shared helper function is readability. It'd make it obvious what the exact differences were.

Patater replied:
I've removed the XEX entry point and its associated tests.

yanesca commented on tests/suites/test_suite_aes.xex.data:
```diff
@@ -0,0 +1,3599 @@
+AES-128-XEX Encrypt NIST XTSTestVectors #1
```
Which NIST document are these defined? (I couldn't find them in NIST SP 800-38E)

gilles-peskine-arm replied:
Please add a few negative tests: lengths that are not a multiple of 16; length=0; AES-192.

yanesca commented on tests/suites/test_suite_aes.xex.data:
```diff
@@ -0,0 +1,3599 @@
+aes_encrypt_xex:"86bd097c948e671deba9cdab9cf7611235eefe5da54ce47eb5c7cbea6c583af7":"0753ba896e5733ed3e1aaa93afeed7af":"779614513f7414929df2bec1d0989dad294c5dc80ec630f56e76b5490ebc43e3":"88fec29008dac2f894bf53364701eb472ed57c542df2ec02942c8b87c201c898"
+
+AES-128-XEX Encrypt NIST XTSTestVectors #240
+aes_encrypt_xex:"b3ede67af12bc4bb90e16
```
I am all for extensive testing, but this seems disproportionately many. We heave 1200 test vectors here and only around 70 for much more heavily used block cipher modes.

gilles-peskine-arm commented on library/gf128mul.c:
```diff
@@ -0,0 +1,78 @@
+    GET_UINT64_LE(a, x, 0);
+    GET_UINT64_LE(b, x, 8);
+
+    ra = (a << 1)  ^ gf128mul_table_bbe[b >> 63];
```
If we're unlucky with memory placement, `gf128mul_table_bbe` may spread over two cache lines and this would leak `b >> 63` to a cache timing attack. I'd feel safer with an approach that is less likely to make different memory loads depending on the value of `b >> 63` and is also unlikely to be compiled to a condition. Maybe this?

    ra = ( a << 1 ) ^ 0x0087 >> ( 8 - ( ( b >> 63 ) << 3 ) );

Patater replied:
I've removed the jump table and used the suggested code. Thanks!

gilles-peskine-arm commented on tests/suites/test_suite_cipher.aes.data:
```diff
@@ -474,6 +474,70 @@ AES-128 CFB - Encrypt and decrypt 32 bytes in multiple parts 1
 depends_on:MBEDTLS_AES_C:MBEDTLS_CIPHER_MODE_CFB
 enc_dec_buf_multipart:MBEDTLS_CIPHER_AES_128_CFB128:128:16:16:-1:16:16:16:16
 
```
Please add negative test cases with 0 bytes and 15 bytes to test error detection. Ideally we should also have a test with 2^24+1 bytes, if this can reasonably be done without allocating that much memory.

Patater replied:
I had a lot of the negative test cases you are describing already. I've added a test for 0 bytes, but already had tests for 15 bytes and 2^24 + 1 bytes. Please let me know if the current set of negative test cases is insufficient.

gilles-peskine-arm commented on tests/suites/test_suite_aes.xts.data:
```diff
@@ -0,0 +1,4337 @@
+AES-128-XTS Encrypt Fail Sector Too Small (by 1 byte)
```
Please add a few more negative tests: length=0; AES-192.

Patater replied:
Fixed

gilles-peskine-arm commented on tests/suites/test_suite_aes.function:
```diff
@@ -151,6 +151,157 @@ exit:
+     * mbedtls_aes_crypt_xts() will prevent any accesses to parameters by
+     * exiting the function early. XXX <-- Hey look, Mr. Codereviewer. What do
+     * you think? Any ideas how to be safe and not have to actually allocate a
+     * 16 MiB buffer nor add NULL checks to mbedtls_aes_crypt_xts()? */
```
I'm fine with the risk of crashing if the length check is not implemented correctly. This is C, it happens sometimes that broken code causes the tests to crash.

On the other hand, at least the context and index should be valid, so that we don't get either a crash because the code tries to dereference them before checking `length` (at least for the context, that's plausible) or a false positive because the code checks those parameters for nullhood before checking `length`.

Due to the same risk of a false positive if the code has additional null checks before the length check, `input` and `output` should be non-null. I propose to make them pointers to a single byte. This makes the behavior on a failed test undefined, but at least there's less of a risk that the test will pass spuriously.

Patater replied:
I've added an initialized context to the test, but did not set any keys in it.

As for input and output being NULL, I would expect a different error returned for NULL checks than for input length. The test is checking for this. I'd rather a NULL get deferenced than some funny values from the stack in the case of an overread.

gilles-peskine-arm commented on tests/suites/test_suite_aes.xex.data:
```diff
@@ -0,0 +1,3599 @@
+aes_encrypt_xex:"86bd097c948e671deba9cdab9cf7611235eefe5da54ce47eb5c7cbea6c583af7":"0753ba896e5733ed3e1aaa93afeed7af":"779614513f7414929df2bec1d0989dad294c5dc80ec630f56e76b5490ebc43e3":"88fec29008dac2f894bf53364701eb472ed57c542df2ec02942c8b87c201c898"
+
+AES-128-XEX Encrypt NIST XTSTestVectors #240
+aes_encrypt_xex:"b3ede67af12bc4bb90e16
```
300 test cases is just a little bit over the top. One test case per length among {16,17,24,31,32,48,512} and per key size would be enough. (I added 24 and 31 because I think half-a-block and block-minus-one are more likely to reveal an off-by-one error somewhere.) This also goes for XTS.

Patater replied:
I deleted all the XEX tests.

I would like to hear also from @yanesca as to which subset of the test vectors we should run with before I remove any.

What would you both think we if removed all the NIST test vectors and use only the IEEE test vectors?

gilles-peskine-arm commented on include/mbedtls/aes.h:
```diff
@@ -88,6 +88,18 @@ typedef struct
 mbedtls_aes_context;
 
+#if defined(MBEDTLS_CIPHER_MODE_XTS)
+/**
```
Please add references for XEX and XTS at the top of the file. (Preexisting: it's also missing references for the other modes of operation.)

Patater replied:
Fixed

Patater commented on library/aes.c:
```diff
@@ -702,6 +720,47 @@ int mbedtls_aes_setkey_dec( mbedtls_aes_context *ctx, const unsigned char *key,
+static int mbedtls_aes_xts_setkey( mbedtls_aes_xts_context *ctx,
+                                   const unsigned char *key,
+                                   unsigned int keybits,
+                                   int (*setkey1)( mbedtls_aes_context *ctx,
```
I decided against that approach because the mode is already being checked earlier and it's not useful to check the mode again; it's better to get the most out of the check we did earlier. I also find this style to be more readable, at both the location where `mbedtls_aes_xts_setkey()` is called and where `setkey1()` is called.

Patater replied:
I've rewritten this to not use function pointers, to not check conditionals unecessarily, and yet still share a lot of code between encrypt and decrypt modes.

Patater commented on include/mbedtls/aes.h:
```diff
@@ -212,6 +281,72 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+ *                     than an AES block in size (16 bytes) or if the length is
+ *                     larger than 2**20 blocks.
+ */
+int mbedtls_aes_crypt_xex( mbedtls_aes_xts_context *ctx,
```
I am skeptical that we need an XEX entry point, since mechanically, it's just a special case of XTS. However, the policy checks we do up front are different. XTS has a limit on the number of blocks, for instance, but (despite what the incorrect comment says) there is no such limit on XEX.

That said, the tests we run are largely duplicate. XEX tests could work with our XTS entry point so long as the number of blocks the test tries to use is not too large (i.e. less than 2 ** 20 blocks, a restriction from NIST SP 800-38E).

yanesca replied:
The reason for the suggested limit is probably that the bit security the security proof guaranties decreases as the size of the data unit rises and at 2**20 blocks the bit security reaches 85.

Typical sector sizes are 0.5kB, and 4kB and 2**20 blocks is 16MB. How likely is it that 16MB is not enough for the foreseeable future?

Patater replied:
I don't know how likely it is to be enough for the foreseeable future. It seems like it'd be enough, but another question is: do we want to enforce this policy in our library? XEX doesn't require it. Only the NIST XTS standard requires this.

yanesca replied:
As I mentioned above the limit is a safe and reasonable choice. If you go above the limit you are not clearly broken, but your security will be below what is considered desirable today.

I've been thinking if we could enforce the limit for now and think it through again if and when somebody complains?

Patater replied:
I'll enforce the limit on XTS.

Are you proposing adding the same limit to the XEX function? How I've currently implemented it, the XEX function is just another entry point to the underlying XTS implementation, but with different policy checks. We could also consider removing the XEX function as the XTS entry point is good enough for all known use cases.

Patater replied:
Additionally, IEEE P1619 defines an XEX construction that uses two keys, unlike the Rogaway paper. So, there is disagreement about what is meant exactly by XEX. This is more reason to remove the XEX API, unless we decide to pick a standard/paper.

Patater replied:
I've removed XEX, since the XTS entry point should be good enough for the applications we know of and XEX isn't standardized.

Patater commented on ChangeLog:
```diff
@@ -2,6 +2,10 @@ mbed TLS ChangeLog (Sorted per branch, date)
 = mbed TLS x.x.x branch released xxxx-xx-xx
 
+Features
+   * Add an implementation of AES-XTS and AES-XTS block cipher modes.
```
Done. I prefer "Add an implementation", but have changed for consistency with the rest of the ChangeLog, as "support for" is more common.

Patater commented on include/mbedtls/aes.h:
```diff
@@ -212,6 +281,72 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+ * \param ctx          The AES XTS context to use for AES XEX operations.
+ * \param mode         The AES operation: #MBEDTLS_AES_ENCRYPT or
+ *                     #MBEDTLS_AES_DECRYPT.
+ * \param length       The of the input data in bytes. The length must be a
```
Fixed

Patater commented on include/mbedtls/aes.h:
```diff
@@ -212,6 +281,72 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+ *                     larger than 2**20 blocks.
+ */
+int mbedtls_aes_crypt_xex( mbedtls_aes_xts_context *ctx,
+                           int mode,
```
Fixed

Patater commented on include/mbedtls/aes.h:
```diff
@@ -212,6 +281,72 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+int mbedtls_aes_crypt_xex( mbedtls_aes_xts_context *ctx,
+                           int mode,
+                           size_t length,
+                           const unsigned char iv[16],
```
Renamed to `tweak`.

Patater commented on include/mbedtls/gf128mul.h:
```diff
@@ -0,0 +1,46 @@
+
+
+
+#endif /* gf128mul.h */
```
I also noticed that the extern "C" was not terminated and fixed this.

Patater commented on library/aes.c:
```diff
@@ -702,6 +720,47 @@ int mbedtls_aes_setkey_dec( mbedtls_aes_context *ctx, const unsigned char *key,
+
+    ret = mbedtls_aes_setkey_enc( &ctx->tweak, key2, key2bits );
+
+    if( ret )
```
Changed to `ret != 0` for consistency with the rest of the library.

Patater commented on library/aes.c:
```diff
@@ -702,6 +720,47 @@ int mbedtls_aes_setkey_dec( mbedtls_aes_context *ctx, const unsigned char *key,
+    const unsigned char *key2 = &key[key1bytes];
+    int ret;
+
+    ret = mbedtls_aes_setkey_enc( &ctx->tweak, key2, key2bits );
```
Fixed

Patater commented on library/cipher.c:
```diff
@@ -440,6 +440,27 @@ int mbedtls_cipher_update( mbedtls_cipher_context_t *ctx, const unsigned char *i
+#if defined(MBEDTLS_CIPHER_MODE_XTS)
+    if( ctx->cipher_info->mode == MBEDTLS_MODE_XTS )
+    {
+        if ( ctx->unprocessed_len ) {
```
Fixed with `> 0`.

Patater commented on library/aes.c:
```diff
@@ -1558,6 +1842,73 @@ int mbedtls_aes_self_test( int verbose )
+        if( mode == MBEDTLS_AES_DECRYPT )
+        {
+            ret = mbedtls_aes_xts_setkey_dec( &ctx_xts, key, 256 );
+            if( ret )
```
Fixed

Patater commented on include/mbedtls/aes.h:
```diff
@@ -141,6 +172,44 @@ int mbedtls_aes_setkey_enc( mbedtls_aes_context *ctx, const unsigned char *key,
+ * \param ctx      The AES XTS context to which the key should be bound.
+ * \param key      The encryption key. This is comprised of the XTS key1
+ *                 concatenated with the XTS key2.
+ * \param keybits  The size of data passed in bits. Valid options are:
```
Fixed

Patater commented on include/mbedtls/aes.h:
```diff
@@ -141,6 +172,44 @@ int mbedtls_aes_setkey_enc( mbedtls_aes_context *ctx, const unsigned char *key,
 
+#if defined(MBEDTLS_CIPHER_MODE_XTS)
+/**
+ * \brief          This function sets the encryption key.
```
Fixed

Patater commented on include/mbedtls/aes.h:
```diff
@@ -212,6 +281,72 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+ *             returns #MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH.
+ *
+ * \param ctx          The AES XTS context to use for AES XTS operations.
+ * \param mode         The AES operation: #MBEDTLS_AES_ENCRYPT or
```
How ciphertext stealing works depends on the mode, so I don't think mode is as redundant with XTS as it is with XEX.

Patater commented on include/mbedtls/aes.h:
```diff
@@ -212,6 +281,72 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+ * \param ctx          The AES XTS context to use for AES XTS operations.
+ * \param mode         The AES operation: #MBEDTLS_AES_ENCRYPT or
+ *                     #MBEDTLS_AES_DECRYPT.
+ * \param length       The length of both an entire data unit and the input
```
Fixed

Patater commented on include/mbedtls/aes.h:
```diff
@@ -212,6 +281,72 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+ * \param length       The length of both an entire data unit and the input
+ *                     data in bytes. The length must be at least 16 bytes. The
+ *                     length does not need to be a multiple of 16 bytes.
+ * \param data_unit    The data unit (commonly a block device sector) address
```
Updated the description.

`data_unit` seems suitable shorthand for the data unit address, just as `sector` would be for the sector address. This is not nominative or anything, just own rationale: I think of CHS as cylinder-head-sector, not cylinder-address-head-address-sector-address, even though the latter is more accurate.

gilles-peskine-arm replied:
When I read the parameter name `data_unit`, I expected it to be a buffer, something of type `uint8_t*` with an associated `size_t data_unit_len`. The original description reinforced this because it started with “the data unit [blah blah]”. It's a lot better now that the description starts with “the address”, but I still find the parameter name misleading, and I'd find `sector` misleading as well if the context was where one would expect to find the sector content. It's different in the context “cylinder-head-sector” because there the words are units, like “degrees-minutes-seconds”.

Patater commented on include/mbedtls/aes.h:
```diff
@@ -141,6 +172,44 @@ int mbedtls_aes_setkey_enc( mbedtls_aes_context *ctx, const unsigned char *key,
+ *                 concatenated with the XTS key2.
+ * \param keybits  The size of data passed in bits. Valid options are:
+ *                 <ul><li>256 bits</li>
+ *                 <li>384 bits</li>
```
Fixed. Per Janos's suggestion, I've added checks for the key length in `mbedtls_aes_xts_setkey()` to catch unsupported key sizes. I've also removed AES-192 from the list of supported key sizes, because they are non-standard.

Patater commented on include/mbedtls/gf128mul.h:
```diff
@@ -0,0 +1,46 @@
+/**
+ * \file gf128mul.h
+ *
+ * \brief Fast multiplication in GF(128)
```
While it is pretty common for crypto modules to have this, I don't think we need that yet for gf128mul. For instance, we don't have ALT for MPI yet.

[I've opened a feature request for this](https://github.com/ARMmbed/mbedtls/issues/1660), as we can add it when we need it without breaking API compatibility.

Patater replied:
Since I've removed the gf128mul module, I've updated the ticket.

Patater commented on library/aes.c:
```diff
@@ -702,6 +720,47 @@ int mbedtls_aes_setkey_dec( mbedtls_aes_context *ctx, const unsigned char *key,
+                                unsigned int keybits)
+{
+    return( mbedtls_aes_xts_setkey( ctx, key, keybits,
+                                    mbedtls_aes_setkey_enc ) );
```
There is a clear benefit to having function pointers here: maintainability (by not repeating one's self), readability (by making clear the exact differences between two implementations, not burying differences between two implementations in a pile of similar looking code).

That said, this is a security sensitive library, so security concerns will trump both maintainability and readability. I'll reconsider if it is possible to meet all three needs: security, maintainability, and readability.

Patater replied:
I've updated the code to try to balance security, maintainability, and readability.

gilles-peskine-arm replied:
I disagree that there's a maintainability benefit to a function pointer here. The part that's repeated is small and could be contained in a single function so that each of `mbedtls_aes_xts_setkey_dec` and `mbedtls_aes_xts_setkey_enc` could be just `{ int ret = common_function(ctx,key,keybits); if (ret == 0) ret = variable_part(); return ret; }`

Patater commented on library/aes.c:
```diff
@@ -983,6 +1042,163 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+                           const unsigned char *input,
+                           unsigned char *output )
+{
+    if( length % 16 )
```
I don't think we always compare with zero explicitly. Consider `mbedtls_aes_crypt_cbc()` for example.

I will compare with zero in the new code, since there seems to be the impression that we always compare with zero explicitly.

Fixed

Patater commented on library/aes.c:
```diff
@@ -983,6 +1042,163 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+        if( mode == MBEDTLS_AES_ENCRYPT )
+        {
+            memcpy( cts_scratch.u8,          (uint8_t*)&inbuf[nblk],              remn );
+            memcpy( cts_scratch.u8 + remn,  ((uint8_t*)&outbuf[nblk - 1]) + remn, 16 - remn );
```
Fixed

Patater commented on library/cipher.c:
```diff
@@ -440,6 +440,27 @@ int mbedtls_cipher_update( mbedtls_cipher_context_t *ctx, const unsigned char *i
+#if defined(MBEDTLS_CIPHER_MODE_XTS)
+    if( ctx->cipher_info->mode == MBEDTLS_MODE_XTS )
+    {
+        if ( ctx->unprocessed_len ) {
```
It's useful for developers of the library to ensure they didn't break something (like the expectations of a certain implementation) when refactoring. I notice we don't have a style of using assertions in the library. Do we have a preferred way of checking for the pre-conditions of a particular implementation? If not, perhaps we can make a note to wrap this in the `REDUNDANT_CHECKS` configuration macros that are Coming Soon™.

yanesca commented on library/aes.c:
```diff
@@ -983,6 +1042,163 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+                           const unsigned char *input,
+                           unsigned char *output )
+{
+    union xts_buf128 {
```
I agree that we should avoid unspecified behaviour, but I would prefer to keep the speed as well. Can't we just use pointers of two different types? Or casting pointers between `uint8_t` and `uint64_t`?

Patater replied:
I've filed an issue https://github.com/ARMmbed/mbedtls/issues/1666 for optimizing all the places we do XOR byte-by-byte. I think a lot of potential speed up is to be had, but I don't want to hold up the XTS review for such an enhancement.

Patater commented on include/mbedtls/gf128mul.h:
```diff
@@ -0,0 +1,46 @@
+ * \param x        the 128-bits number you want to multiply
+ * \param r        result
+ */
+void mbedtls_gf128mul_x_ble(mbedtls_be128 r, const mbedtls_be128 x);
```
I've inlined the function and removed the gf128mul module. It's only used from XTS currently, and it is questionable if its useful outside of XTS. We can move it out to a module when we need it.

gilles-peskine-arm replied:
@yanesca I was thinking of Camellia-XTS (which we don't support now, but it would be natural to have it). As for GCM, I didn't realize that this function wouldn't be a useful building block for GHASH, but I still think it would make sense to have a module for GF(2^128) operations. Ok, this can wait.

Patater commented on library/aes.c:
```diff
@@ -983,6 +1042,163 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+    if( mode == MBEDTLS_AES_DECRYPT && remn )
+    {
+        if( nblk == 1 )
+            goto decrypt_only_one_full_block;
```
Fixed by rewrite

Patater commented on library/aes.c:
```diff
@@ -983,6 +1042,163 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+    if( length % 16 )
+        return( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+
+    return( mbedtls_aes_crypt_xts( ctx, mode, length,
```
Fixed. I've removed the XEX entry point.

Patater commented on library/CMakeLists.txt:
```diff
@@ -28,6 +28,7 @@ set(src_crypto
     entropy_poll.c
     error.c
     gcm.c
+    gf128mul.c
```
Removed gf128mul module, so not needed in any of these places anymore

Patater commented on library/aes.c:
```diff
@@ -983,6 +1042,163 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+/*
+ * AES-XEX buffer encryption/decryption
+ */
+int mbedtls_aes_crypt_xex( mbedtls_aes_xts_context *ctx,
```
I've deleted this function, so this concern is no longer relevant.

Patater commented on library/aes.c:
```diff
@@ -983,6 +1042,163 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+    if( mode == MBEDTLS_AES_DECRYPT && remn )
+    {
+        if( nblk == 1 )
+            goto decrypt_only_one_full_block;
```
The code has been rewritten to avoid goto.

Patater commented on library/cipher_wrap.c:
```diff
@@ -337,6 +367,89 @@ static const mbedtls_cipher_info_t aes_256_ctr_info = {
+{
+    mbedtls_aes_xts_context *xts_ctx = mbedtls_calloc( 1, sizeof( *xts_ctx ) );
+
+    if( xts_ctx )
```
Fixed

Patater commented on library/cipher_wrap.c:
```diff
@@ -337,6 +367,89 @@ static const mbedtls_cipher_info_t aes_256_ctr_info = {
+{
+    mbedtls_aes_xts_context *xts_ctx = ctx;
+
+    if ( !xts_ctx )
```
Fixed

Patater commented:
I've rebased the commits, addressing your feedback along the way. Please give another round of review. Reviewing commit by commit should be easiest, but you may also want to diff against [the previous branch you reviewed](https://github.com/Patater/mbedtls/tree/dev/patater/xts-r1). Thanks very much!

gilles-peskine-arm commented on include/mbedtls/aes.h:
```diff
@@ -212,6 +281,72 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+ * \param output       The buffer holding the output data.
+ *
+ * \return             \c 0 on success.
+ * \return             #MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH if length smaller
```
I prefer not to emphasize the block cipher block size in XTS documentation because XTS is mostly used to encrypt storage, thus users are likely to have storage in mind more than cryptography, and in the context of storage a block has a different meaning. But that's minor.

Patater commented:
I've added two more commits, adding negative tests for key sizes and fixing the description of gf128mul.

Patater commented on ChangeLog:
```diff
@@ -2,6 +2,10 @@ mbed TLS ChangeLog (Sorted per branch, date)
 = mbed TLS x.x.x branch released xxxx-xx-xx
 
+Features
+   * Add support for the AES-XEX and AES-XTS (NIST 800-38E) block cipher modes.
```
I don't believe we want to. I'll remove this.

Fixed

Patater commented on include/mbedtls/aes.h:
```diff
@@ -141,6 +176,44 @@ int mbedtls_aes_setkey_enc( mbedtls_aes_context *ctx, const unsigned char *key,
 
+#if defined(MBEDTLS_CIPHER_MODE_XTS)
+/**
+ * \brief          This function prepares and XTS context for encryption and
```
Fixed

Patater commented on library/aes.c:
```diff
@@ -702,6 +716,78 @@ int mbedtls_aes_setkey_dec( mbedtls_aes_context *ctx, const unsigned char *key,
+
+    ret = mbedtls_aes_xts_decode_keys( key, keybits, &key1, &key1bits,
+                                       &key2, &key2bits );
+    if (ret != 0 )
```
Fixed

Patater commented on library/aes.c:
```diff
@@ -702,6 +716,78 @@ int mbedtls_aes_setkey_dec( mbedtls_aes_context *ctx, const unsigned char *key,
+
+    ret = mbedtls_aes_xts_decode_keys( key, keybits, &key1, &key1bits,
+                                       &key2, &key2bits );
+    if (ret != 0 )
```
Fixed

Patater commented on library/aes.c:
```diff
@@ -983,6 +1069,164 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+    unsigned char tmp[16];
+
+    /* Sectors must be at least 16 bytes. */
+    if( length < 16)
```
Fixed

Patater commented on library/aes.c:
```diff
@@ -983,6 +1069,164 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+        return MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH;
+
+    /* NIST SP 80-38E disallows data units larger than 2**20 blocks. */
+    if( (blocks > 0x100000) || (blocks == 0x100000 && leftover > 0))
```
Fixed

Patater commented:
I've removed the NIST test vectors, leaving only the IEEE P1619 vectors present, plus the negative tests.

gilles-peskine-arm commented on include/mbedtls/aes.h:
```diff
@@ -212,6 +281,72 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+ *             Length must be a multiple of the AES block size (16 bytes).
+ *
+ * \param ctx          The AES XTS context to use for AES XEX operations.
+ * \param mode         The AES operation: #MBEDTLS_AES_ENCRYPT or
```
Ah, right. You can't tell from the context. Pity. This is indeed a pre-existing defect with ECB and CBC. This is unfortunate but I'm ok with not going beyond what was done for CBC so I retract my objection.

Patater commented:
I've rebased to address changes. See previously reviewed branch at https://github.com/Patater/mbedtls/tree/dev/patater/xts-r2

Assuming you have Patater added as a remote, diff manually via
`git diff Patater/dev/patater/xts-r2..Patater/dev/patater/xts`

Convenient diff output at https://gist.github.com/Patater/04c2175b15091edb5056740dc109bcc3

Patater commented on ChangeLog:
```diff
@@ -2,6 +2,10 @@ mbed TLS ChangeLog (Sorted per branch, date)
 = mbed TLS x.x.x branch released xxxx-xx-xx
 
+Features
+   * Add support for the AES-XTS (NIST 800-38E) block cipher mode.
```
Updated

Patater commented on ChangeLog:
```diff
@@ -2,6 +2,10 @@ mbed TLS ChangeLog (Sorted per branch, date)
 = mbed TLS x.x.x branch released xxxx-xx-xx
 
+Features
+   * Add support for the AES-XTS (NIST 800-38E) block cipher mode.
```
Removed NIST reference.

Patater commented on include/mbedtls/aes.h:
```diff
@@ -13,6 +13,9 @@
  *          ciphers</em>.
+ *
+ *          The AEX-XTS block mode is standardized by NIST 800-38E
+ *          <https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-38e.pdf>.
```
Fixed

Patater commented on include/mbedtls/aes.h:
```diff
@@ -13,6 +13,9 @@
  *          ciphers</em>.
+ *
+ *          The AEX-XTS block mode is standardized by NIST 800-38E
+ *          <https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-38e.pdf>.
```
Now referring to both documents, since you need both as an implementer

Patater commented on include/mbedtls/aes.h:
```diff
@@ -212,6 +285,49 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+ *             operation for an entire XTS data unit.
+ *
+ *             AES-XTS encrypts or decrypts blocks based on their location as
+ *             defined by a data unit number. These must be provided by \p
```
Changed to "The data unit number must be provided..."

Patater commented on include/mbedtls/aes.h:
```diff
@@ -212,6 +285,49 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+ *                     is typically the index of the block device sector that
+ *                     contains the data.
+ * \param input        The buffer holding the input data (which is an entire
+ *                     data unit). \p input must be at least \p length bytes
```
I can understand how some users would be confused by "at least" implying that we might do more, but it is technically correct.

Now we, instead of telling users what they need to do, say what the function does and it's up to the user to provide a buffer capable of meeting the stated needs of the function.

Patater commented on include/mbedtls/config.h:
```diff
@@ -495,6 +495,14 @@
+ * \def MBEDTLS_CIPHER_MODE_XTS
+ *
+ * Enable Xor-encrypt-xor with ciphertext stealing mode (XTS) for symmetric
+ * ciphers.
```
Added

Patater replied:
Added

Patater commented on library/aes.c:
```diff
@@ -983,6 +1069,166 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+        return MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH;
+
+    /* NIST SP 80-38E disallows data units larger than 2**20 blocks. */
+    if( ( blocks > 0x100000 ) || ( blocks == 0x100000 && leftover > 0 ) )
```
Agreed. I've updated this.

Patater commented on library/aes.c:
```diff
@@ -983,6 +1069,166 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+        return MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH;
+
+    /* Compute the tweak. */
+    memcpy( tweak, data_unit, sizeof( tweak ) );
```
Fixed

Patater commented on library/aes.c:
```diff
@@ -983,6 +1069,166 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+    size_t blocks = length / 16;
+    size_t leftover = length % 16;
+    unsigned char tweak[16];
+    unsigned char tweak_tmp[16];
```
Sounds good to me. Updated.

Patater commented on library/aes.c:
```diff
@@ -983,6 +1069,166 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+
+        /* Copy ciphertext bytes from previous block to our output for each
+         * byte of cyphertext we won't steal. At the same time, the remainder
+         * of input for this final round (since the loop bounds are the same).
```
Added the missing word "copy" in the right spot. Added a few more articles as well.

Patater commented on programs/test/benchmark.c:
```diff
@@ -94,8 +94,8 @@ int main( void )
-    "aes_cbc, aes_gcm, aes_ccm, aes_cmac, des3_cmac,\n"                 \
-    "havege, ctr_drbg, hmac_drbg\n"                                     \
+    "aes_cbc, aes_gcm, aes_ccm, aes_cmac, aes_xts,\n"                   \
+    "des3_cmac, havege, ctr_drbg, hmac_drbg\n"                          \
```
Fixed

Patater commented on programs/test/benchmark.c:
```diff
@@ -419,6 +421,29 @@ int main( int argc, char *argv[] )
+
+            memset( buf, 0, sizeof( buf ) );
+            memset( tmp, 0, sizeof( tmp ) );
+            mbedtls_aes_xts_setkey_enc( &ctx, tmp, keysize * 2);
```
Fixed

Patater commented on include/mbedtls/cipher_internal.h:
```diff
@@ -71,6 +71,13 @@ struct mbedtls_cipher_base_t
 #endif
 
+#if defined(MBEDTLS_CIPHER_MODE_XTS)
+    /** Encrypt using XTS */
```
Changed to prefer correctness over consistency

Patater commented:
I've fixed up the commits with feedback, and added a couple new commits.

See previously reviewed branch at https://github.com/Patater/mbedtls/tree/dev/patater/xts-r3

Assuming you have Patater added as a remote, diff manually via
git diff Patater/dev/patater/xts-r3..Patater/dev/patater/xts

Convenient diff at: https://gist.github.com/Patater/b701db602740b9352eaf7f1ef8b6d271

gilles-peskine-arm commented on include/mbedtls/aes.h:
```diff
@@ -13,6 +13,11 @@
+ *
+ *          The AES-XTS block mode is standardized by NIST SP 800-38E
+ *          <https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-38e.pdf>
+ *          and the described in detail by IEEE P1619
```
“the” → “is”, “by” → “in”

Patater replied:
"by" is OK here. I'll get rid of the extra "the".

Patater replied:
"by" is OK here. I've gotten rid of the extra "the".

Patater commented:
See previously reviewed branch at https://github.com/Patater/mbedtls/tree/dev/patater/xts-r4

Assuming you have Patater added as a remote, diff manually via
git diff Patater/dev/patater/xts-r4..Patater/dev/patater/xts

https://gist.github.com/Patater/5962dfa5a5ca1c4d4e58a22644518917

Patater commented:
Currently failing Debian 9 i386 CMake ASan "make test". Needs fixing before merging.

Patater commented:
I've added a new commit for the fix for the ASan failure on i386. After it is reviewed, and CI passes, this should be good to go.

Patater commented:
timing-suite on FreeBSD failed, as it often tends to, known issue.

sbutcher-arm commented:
@Patater - Can you please arrange for the branch conflicts to be fixed?

Patater commented:
Rebased on latest development branch.

Old branch at https://github.com/Patater/mbedtls/tree/dev/patater/xts-r6
New branch at https://github.com/Patater/mbedtls/tree/dev/patater/xts

Diff (which also contains upstream changes) at https://gist.github.com/Patater/8ea307eb7d1fce28002e479f41a952af

The only changes between this version of the PR and the previous are:
- New things from upstream
- ChangeLog entry moved to new x.xx.x version
- Added xts alongside ARIA in benchmark.c (to resolve conflicts)

Patater commented:
@yanesca To catch the ChangeLog issue, I looked at the context around where the entry was added when diffing the branch before rebase and the branch after rebase. Git thought it was smart and auto-resolved the addition of the entry by adding it in its original place below the other features. However, since those features had been released in the mean time, which git has no way of knowing, the auto-resolution was incorrect.

Patater commented:
Looks like cipher wrapper conflicts weren't resolved correctly. Fixes coming soon.

Patater commented:
Fixes now included in this PR.

Old branch https://github.com/Patater/mbedtls/tree/dev/patater/xts-r6
New branch https://github.com/Patater/mbedtls/tree/dev/patater/xts

Diff (adds XTS to aria cipher info structs) https://gist.github.com/Patater/653ff4d3cdf94a0caff0dba875dd6c7c

Note that OFB is expected to merge first and may conflict with this PR in a non-trivial way, so we may need yet another rebase.

Patater commented:
Jenkins CI failure is just the known issue of timing-test on FreeBSD.

```
      Start 57: timing-suite

57/62 Test #57: timing-suite .....................***Failed    3.30 sec
```

Patater commented:
Rebased to resolve conflicts.

Old branch https://github.com/Patater/mbedtls/tree/dev/patater/xts-r8
New branch https://github.com/Patater/mbedtls/tree/dev/patater/xts

Diff between r8 and current version (includes new things merged upstream) https://gist.github.com/Patater/9da8163d810c3723414d2ae3d50e7965

The only changes between this version of the PR and the previous are:
- New things from upstream
- ChangeLog entry moved to upstream's x.x.x version Feature section
- Added XTS alongside OFB in `tests/Makefile`, `test_suite_cipher.aes.data`, `cipher_wrap.c`, and `cipher.h`

yanesca commented:
Are there any expected rebases left?

Patater commented:
No.

Patater commented:
@yanesca I ran `git commit-tree xts-r9^{tree} -p xts-r8 -p origin/development -m "Mergediff"` and a `git show` on the resulting hash to create and view a detached (not on any branches) merge commit between r9 and r8 that includes all the conflict resolution I had to do when moving the base to the latest development branch plus all the changes made in r9 since r8 (none really in this case, other than conflict resolution), hiding all changes made upstream in the new base that didn't conflict.

Patater commented on include/mbedtls/config.h:
```diff
@@ -501,6 +501,13 @@
 #define MBEDTLS_CIPHER_MODE_CBC
 
+/**
+ * \def MBEDTLS_CIPHER_MODE_XTS
```
Fixed

Patater commented:
retest
